pyOCD/coresight/ap.py

```diff
@@ -0,0 +1,347 @@
+        if self.has_rom_table and bus_accessible:
+            self.init_rom_table()
+
+    def init_rom_table(self):
+        self.rom_table = ROMTable(self)
+        self.rom_table.init()
+
+    def read_reg(self, addr, now=True):
+        return self.dp.readAP((self.ap_num << APSEL_SHIFT) | addr, now)
+
+    def write_reg(self, addr, data):
+        self.dp.writeAP((self.ap_num << APSEL_SHIFT) | addr, data)
+
+class MEM_AP(AccessPort):
+    def __init__(self, dp, ap_num):
+        super(MEM_AP, self).__init__(dp, ap_num)
+        self.auto_increment_page_size = 4
+
+    ## @brief Write a single memory location.
+    #
+    # By default the transfer size is a word
+    def writeMemory(self, addr, data, transfer_size=32):
+        num = self.dp.next_access_number
+        if LOG_DAP:
+            self.logger.info("writeMem:%06d (addr=0x%08x, size=%d) = 0x%08x {", num, addr, transfer_size, data)
+        self.write_reg(AP_REG['CSW'], CSW_VALUE | TRANSFER_SIZE[transfer_size])
+        if transfer_size == 8:
+            data = data << ((addr & 0x03) << 3)
+        elif transfer_size == 16:
+            data = data << ((addr & 0x02) << 3)
+
+        try:
+            self.write_reg(AP_REG['TAR'], addr)
+            self.write_reg(AP_REG['DRW'], data)
+        except DAPAccess.Error as error:
+            self._handle_error(error, num)
+            raise
+        if LOG_DAP:
+            self.logger.info("writeMem:%06d }", num)
+
+    ## @brief Read a memory location.
+    #
+    # By default, a word will be read.
+    def readMemory(self, addr, transfer_size=32, now=True):
+        num = self.dp.next_access_number
+        if LOG_DAP:
+            self.logger.info("readMem:%06d (addr=0x%08x, size=%d) {", num, addr, transfer_size)
+        res = None
+        try:
+            self.write_reg(AP_REG['CSW'], CSW_VALUE |
+                         TRANSFER_SIZE[transfer_size])
+            self.write_reg(AP_REG['TAR'], addr)
+            result_cb = self.read_reg(AP_REG['DRW'], now=False)
+        except DAPAccess.Error as error:
+            self._handle_error(error, num)
+            raise
+
+        def readMemCb():
+            try:
+                res = result_cb()
+                if transfer_size == 8:
+                    res = (res >> ((addr & 0x03) << 3) & 0xff)
+                elif transfer_size == 16:
+                    res = (res >> ((addr & 0x02) << 3) & 0xffff)
+                if LOG_DAP:
+                    self.logger.info("readMem:%06d %s(addr=0x%08x, size=%d) -> 0x%08x }", num, "" if now else "...", addr, transfer_size, res)
+            except DAPAccess.Error as error:
+                self._handle_error(error, num)
+                raise
+            return res
+
+        if now:
+            result = readMemCb()
+            return result
+        else:
+            return readMemCb
+
+    # write aligned word ("data" are words)
+    def writeBlock32(self, addr, data):
```

Comment:
this should probably be made private since it can only be used correctly if auto_increment_page_size is known.